rasa/utils/tensorflow/layers.py

```diff
@@ -198,24 +188,150 @@ class DenseWithSparseWeights(tf.keras.layers.Dense):
+            bias_regularizer: Regularizer function applied to the bias vector.
+            activity_regularizer: Regularizer function applied to
+                the output of the layer (its "activation")..
+            kernel_constraint: Constraint function applied to
+                the `kernel` weights matrix.
+            bias_constraint: Constraint function applied to the bias vector.
+        """
         super().__init__(**kwargs)
-        self.sparsity = sparsity
+
+        if density < 0.0 or density > 1.0:
+            raise TFLayerConfigException("Layer density must be in [0, 1].")
+
+        self.density = density
 
     def build(self, input_shape: tf.TensorShape) -> None:
+        """Prepares the kernel mask.
+
+        Args:
+            input_shape: Shape of the inputs to this layer
+        """
         super().build(input_shape)
-        # create random mask to set fraction of the `kernel` weights to zero
-        kernel_mask = tf.random.uniform(tf.shape(self.kernel), 0, 1)
-        kernel_mask = tf.cast(
-            tf.greater_equal(kernel_mask, self.sparsity), self.kernel.dtype
+
+        if self.density == 1.0:
+            self.kernel_mask = None
+            return
+
+        kernel_shape = tf.shape(self.kernel)
+        num_rows = kernel_shape[0].numpy()
+        num_cols = kernel_shape[1].numpy()
+
+        # Construct mask with given density and guarantee that every output is
+        # connected to at least one input
+
+        min_num_connected = max(num_rows, num_cols)
+        num_extra_connections = max(
+            tf.cast(
+                tf.math.ceil(
+                    self.density * num_rows * num_cols
+                    - (1.0 - self.density) * min_num_connected
+                ),
+                tf.int32,
+            ),
+            0,
+        )
+
+        kernel_mask = self._minimal_mask(
+            num_rows, num_cols, dtype=self.kernel.dtype
+        ) + self._random_mask(
+            num_rows, num_cols, num_extra_connections, dtype=self.kernel.dtype
         )
+
+        # We might accidently have added a random connection on top of
+        # a fixed connection
+        kernel_mask = tf.clip_by_value(kernel_mask, 0, 1)
+
         self.kernel_mask = tf.Variable(
             initial_value=kernel_mask, trainable=False, name="kernel_mask"
         )
 
+    def _random_mask(
+        self, num_rows: int, num_cols: int, num_ones: int, dtype: tf.dtypes.DType
+    ) -> tf.Tensor:
+        """Creates a random matrix with `num_ones` 1s and 0s otherwise.
+
+        Args:
+            num_rows: Number of rows in the mask
+            num_cols: Number of columns in the mask
+            num_ones: Number of ones in the matrix
+            dtype: Type of the values
+
+        Returns:
+            A random mask matrix
+        """
+        mask = tf.random.uniform(tf.shape(self.kernel), 0, 1)
+        mask = tf.cast(tf.less_equal(mask, self.density), self.kernel.dtype)
```

Comment:
I think the convention is to use `<` for upper bounds
```suggestion
        mask = tf.cast(tf.math.less(mask, self.density), self.kernel.dtype)
```